### ISSUE TITLE
Time Series Model

### ISSUE THREAD

**Issue by aash1999** [pranavdhawann/spring-2026-group4#39]
### Summary

Introduce a **generic, configurable PyTorch-based time series model** (e.g., LSTM or similar) under `src/models`.
This model should serve as a **reusable building block** that can be easily extended or composed for multiple time series use cases.

---

## 📄 Description

We need a **flexible and modular PyTorch time series model implementation** that follows best practices and avoids hard-coded configurations.
All model behavior and architecture parameters should be configurable via a **parameter dictionary**, with sensible defaults applied when parameters are not explicitly provided.

The model should be easy to:

* Reuse across experiments
* Extend with additional layers or architectures
* Integrate into standard PyTorch training loops

---

### Model Design

* Implement the model as a class that **inherits from `torch.nn.Module`**
* The class **must define**:

  * `__init__(self, config: dict)`
  * `forward(self, x, **kwargs)`
* **No hard-coded parameters**

  * All hyperparameters (e.g., `input_size`, `hidden_size`, `num_layers`, `dropout`, `output_size`, etc.) must be configurable
  * Parameters should be passed via a **dictionary**
  * If a parameter is missing, a **default value must be used**

### Code Quality

* Follow **PyTorch best practices and conventions**
* Keep the code **readable, modular, and well-documented**
* Avoid assumptions about batch size or sequence length in `forward`

---

## 🧪 Reference Example (Style Only)

The following example is **for structure and style reference only**:

```python
import torch

class TinyModel(torch.nn.Module):
    def __init__(self):
        super(TinyModel, self).__init__()
        self.linear1 = torch.nn.Linear(100, 200)
        self.activation = torch.nn.ReLU()
        self.linear2 = torch.nn.Linear(200, 10)

    def forward(self, x):
        x = self.linear1(x)
        x = self.activation(x)
        x = self.linear2(x)
        return x
```

> ⚠️ **Important Note:**
> The actual implementation **must not hard-code values** as shown above.
> All parameters must be configurable via a dictionary with defaults.

---

## 🧱 Example Configuration Dictionary

```python
config = {
    "input_size": 16,
    "hidden_size": 64,
    "num_layers": 2,
    "dropout": 0.1,
    "output_size": 1,
    "bidirectional": False
}
```

If a key is missing, the model should automatically fall back to a default value.

---

## ▶️ Example Driver Code (Required)

The same file **must include example usage / driver code**, such as:

```python
if __name__ == "__main__":
    import torch

    config = {
        "input_size": 16,
        "hidden_size": 64,
        "num_layers": 2,
        "output_size": 1
    }

    model = TimeSeriesModel(config)
    x = torch.randn(8, 20, 16)  # (batch_size, seq_len, input_size)
    y = model(x)

    print(y.shape)
```

This example is intended to:

* Demonstrate correct model initialization
* Validate the forward pass
* Serve as quick reference for contributors

---

## 📦 Expected Outcome

* A new file under `src/models/`

  * Example: `lstm.py`, `timeseries_model.py`
* A clean, reusable PyTorch time series model class
* Configurable architecture via a parameter dictionary
* Built-in defaults for missing configuration values
* Example driver code in the same file demonstrating usage

---

## ✅ Acceptance Criteria

* [ ] Model class inherits from `torch.nn.Module`
* [ ] `__init__` and `forward` methods are implemented
* [ ] No hard-coded model dimensions or hyperparameters
* [ ] Model configuration is passed via a dictionary with defaults
* [ ] Input tensors are handled dynamically in `forward`
* [ ] Code follows PyTorch best practices
* [ ] Example driver code is included and runnable
* [ ] Model output shape is consistent with `output_size`

---

**Comment by aash1999** [comment]
[File to LSTM](https://github.com/pranavdhawann/spring-2026-group4/blob/dev/src/models/timeseries_lstm.py)
